File: 5-longest-palindromic-substring/manachers-algorithm/manacher.cpp
#include <string>

class Solution
{
    public:
        string longestPalindrome(string s)
        {
            if (s.length() == 1)
            {
                return s;
            }
            
            int N = 2 * s.length() + 1;

            // LPS Array
            int LPS[N];
            LPS[0] = 0;
            LPS[1] = 1;

            int l = 0; // Current Left Position
            int c = 1; // Center Position
            int r = 2; // Center Right Position

            // Other Variables
            int expand = -1;
            int diff = -1;
            int maxLPSLength = 0;
            int maxLPSPosition = 0;
            int start = -1;
            
            for (int i = 2; i < N; i++)
            {
                // Get Current Left Position
                l = 2 * c - i;

                // Reset expand
                expand = 0;

                // If current right position is within center right
                diff = r - i;
                if (diff >= 0)
                {
                    if (LPS[l] < diff)
                    {
                        LPS[i] = LPS[l];
                    }
                    else if ((LPS[l] == diff) && (r == (N - 1)))
                    {
                        LPS[i] = LPS[l];
                    }
                    else if ((LPS[l] == diff) && (r < (N - 1)))
                    {
                        LPS[i] = LPS[l];

                        //expansion require
                        expand = 1;
                    }
                    else if (LPS[l] > diff)
                    {
                        LPS[i] = diff;

                        //epxnaion required
                        expand = 1;
                    }
                }
                else
                {
                    LPS[i] = 0;

                    //expansion required
                    expand = 1;
                }

                if (expand)
                {
                    //Expand palindrom centered at i
                    while (((i + LPS[i]) < N && (i - LPS[i]) > 0) && (((i + LPS[i] + 1) % 2 == 0) || (s[(i + LPS[i] + 1) / 2] == s[(i - LPS[i] - 1) / 2])))
                    {
                        LPS[i]++;
                    }
                }

                // Update max LPS length
                if (LPS[i] > maxLPSLength)
                {
                    maxLPSLength = LPS[i];
                    maxLPSPosition = i;
                }

                // Update center and right positions
                if (i + LPS[i] > r)
                {
                    c = i;
                    r = i + LPS[i];
                }
            }

            start = (maxLPSPosition - maxLPSLength) / 2;

            return s.substr(start, maxLPSLength);
        }
};

File: 5-longest-palindromic-substring/manachers-algorithm/manacher.rs
use std::cmp;

impl Solution
{
    pub fn longest_palindrome(s: String) -> String
    {
        if (s.len() == 1)
        {
            return s;
        }
        
        let N: usize = 2 * s.len() + 1;
        let mut LPS: [i32; 2005] = [0; 2005];
        LPS[0] = 0;
        LPS[1] = 1;

        let mut l: i32 = 0; // Current Left Position
        let mut c: i32 = 1; // Center Position
        let mut r: i32 = 2; // Center Right Position

        // Other Variables
        let mut maxLPSLength: i32 = 0;
        let mut maxLPSPosition: i32 = 0;

        for i in 2..N
        {
            // Get Current Left Position
            l = 2 * c - l;

            let expand: bool = false;

            // If current right position is within center right
            let j: i32 = i as i32;
            let k: usize = l as usize;
            let q: i32 = N as i32;

            let diff: i32 = r - j;
            if (diff >= 0)
            {
                if (LPS[k] < diff)
                {
                    LPS[i] = LPS[k];
                }
                else if ((LPS[k] == diff) && (r == (q - 1)))
                {
                    LPS[i] = LPS[k];
                }
                else if ((LPS[k] == diff) && (r < (q - 1)))
                {
                    LPS[i] = LPS[k];

                    //expansion required
                    expand = true;
                }
                else
                {
                    LPS[i] = diff;

                    //epxnaion required
                    expand = true;
                }
            }
            else
            {
                LPS[i] = 0;
                expand = true;
            }

            if (expand)
            {
                // Expand palindrome centered at i
                while (true)
                {
                    let u: usize = ((j + LPS[i] + 1) / 2) as usize;
                    let v: usize = ((j - LPS[i] - 1) / 2) as usize;

                    if (u >= s.len())
                    {
                        break;
                    }
                    else if (v <= 0)
                    {
                        break;
                    }
                    else if (((j + LPS[i] + 1) % 2 != 0) && (s[u] != s[v]))
                    {
                        break;
                    }
                    else
                    {
                        LPS[i] += 1;
                    }
                }
            }

            // Update max LPS length
            if (LPS[i] > maxLPSLength)
            {
                maxLPSLength = LPS[i];
                maxLPSPosition = i;
            }

            // Update center and right positions
            if (i + LPS[i] > r)
            {
                c = i;
                r = i + LPS[i];
            }
        }

        return s;
    }
}